app/src/events/position_state_changed.c

```diff
@@ -7,4 +7,15 @@
 #include <kernel.h>
 #include <zmk/events/position_state_changed.h>
 
-ZMK_EVENT_IMPL(zmk_position_state_changed);
\ No newline at end of file
+ZMK_EVENT_IMPL(zmk_position_state_changed);
+
+uint32_t zmk_last_event_trace_id = 0;
+uint32_t zmk_event_trace_ids[ZMK_KEYMAP_LEN] = {0};
+
+uint32_t zmk_event_trace_id(uint32_t position, bool generate) {
```

Comment:
Maybe add "get" somewhere in the name? Having a verb somewhere in the name of a function usually makes it easier to understand what it's doing (unless you mean for "trace" to be the verb, but I'm assuming not given that it's used in variable names too).

Also, I'm generally not a fan of boolean parameters since in code like
```c
zmk_event_trace_id(position, TRUE);
```
you have no idea what the 2nd parameter does without looking at the definition of the function. That said, it is convenient for the couple places this is being called, though if it starts getting used elsewhere you may run into the same readability problem.

Consider something like this, which is more verbose but tells you exactly what it's doing:
```c
uint32_t zmk_event_trace_id_get(uint32_t position) {
  return zmk_event_trace_ids[position];
}
uint32_t zmk_event_trace_id_create(uint32_t position) {
  zmk_last_event_trace_id++;
  zmk_event_trace_ids[position] = zmk_last_event_trace_id;
  return zmk_event_trace_id_get(position);
}

...
event = {
  .trace_id = pressed ? zmk_event_trace_id_create(position) : zmk_event_trace_id_get(position),
};
```

Alternatively:
```c
uint32_t zmk_event_trace_id_get(uint32_t position) {
  return zmk_event_trace_ids[position];
}
void zmk_event_trace_id_update(uint32_t position) {
  zmk_last_event_trace_id++;
  zmk_event_trace_ids[position] = zmk_last_event_trace_id;
}

...
if (pressed) {
  zmk_event_trace_id_update(position);
}
event = {
  .trace_id = zmk_event_trace_id_get(position),
};
```

Reply:
Thanks for your extensive reply! I've added a commit that implements the create/get version.  I like how the intent is clearer when calling with a hardcoded value. 

On the other hand this solution exposes an implementation detail of trace_ids to the caller, by making the caller responsible for asking for a new trace_id or retrieving an existing trace_id. Having this decision within the trace_id method puts the responsibility in a better place in my opinion.

The update/get variety may be the best middle ground between the two. 

I see arguments for each option, I'm not quite sure what the best one is yet.

Reply:
> On the other hand this solution exposes an implementation detail of trace_ids to the caller, by making the caller responsible for deciding if a new trace_id should be created or an existing one retrieved.

I'd argue that the caller is still responsible for deciding that in the version that takes a boolean parameter. You just one have one function that does one of two things based on a parameter instead of two functions that each do one thing. 

If you want to hide away the fact that new trace IDs get generated on key press, then renaming the parameter to something like `is_press` (like you mentioned) would be better, though it would still be difficult to understand if any later code ends up needing to pass just `TRUE` or `FALSE` for that parameter. You could also rename the "update" function to something like `zmk_event_trace_id_notify_pressed(position)` (though then it's not super clear if you're notifying the trace ID system of a press, as intended, or if you want to send a notification). Naming things is hard.

Reply:
I've decided to go with `zmk_get_event_trace_id(uint32_t position, bool pressed)`, as I think it's the clearest code for now. I don't expect other call sites that will use literal booleans, and if that will be the case we can always refactor this method.